doc/userguide/rules/rules-internals.rst

```diff
@@ -0,0 +1,151 @@
+
+#. :ref:`Action <action>`
+#. Usage of :ref:`flowbits`
+#. Usage of :ref:`flowint`
+#. Usage of flowvar
+#. Usage of pktvar
+#. Usage of hostbits
+#. Usage of ippair
+#. :ref:`Priority <priority>`
+
+In this order. Once signatures are ordered, they are attributed a unique
+internal id (``Signature::num``) which symbolizes their priority (the lower the
+``num``, the higher the priority). This could mean that a rule with keyword
+defined priority 1 could have lower priority than another rule, if the other has
+flowbits set and an action with higher priority, for instance.
+
+Example
+~~~~~~~
+
+Consider the rules bellow:
+
+.. container:: example-rule
+
+    pass udp any any -> any 6081 (sid:1; gid:10000003;)
+
+.. container:: example-rule
+
+    pass tls $HOME_NET any -> $EXTERNAL_NET any (tls.sni; content:"some.\
+    content"; startswith; nocase; endswith; msg:"matching TLS allowed";\
+    priority:1; flow:to_server, established; sid:2; rev:1; gid:10000003;)
+
+.. container:: example-rule
+
+    drop tls $HOME_NET any -> $EXTERNAL_NET any (msg:"not matching any TLS\
+    allowed"; priority:1; flow:to_server, established; sid:3; rev:1; gid:\
+    10000003;)
+
+.. container:: example-rule
+
+    pass udp any any -> any 6081 (sid:4; gid:10000003;)
+
+.. container:: example-rule
+
+    pass tls $HOME_NET any -> $EXTERNAL_NET any (tls.sni; content:"some.\
+    content"; startswith; nocase; endswith; msg:"matching TLS allowed";\
+    priority:1; flow:to_server, established; sid:5; rev:1; gid:10000003;)
+
+.. container:: example-rule
+
+    pass tls $HOME_NET any -> $EXTERNAL_NET any (tls.sni; content:"some.\
+    content"; startswith; nocase; endswith; msg:"matching TLS allowed";\
+    priority:1; flow:to_server, established; content:"hypothetical"; sid:6;\
+    rev:1; gid:10000003;)
+
+With the default `action order configuration <suricata-yaml-action-order>`_
+(*pass*, *drop*, *reject*, *alert*), after loading and sorting, their
+internal IDs would be:
+
++--------------------+-------------------+
+| Signature ID (sid) | Internal ID (num) |
++====================+===================+
+|  1                 |  3                |
++--------------------+-------------------+
+|  2                 |  0                |
++--------------------+-------------------+
+|  3                 |  5                |
++--------------------+-------------------+
+|  4                 |  4                |
++--------------------+-------------------+
+|  5                 |  1                |
++--------------------+-------------------+
+|  6                 |  2                |
++--------------------+-------------------+
+
+
+Inspection Flow
+---------------
+
+Once it is time to inspect network traffic against the loaded rules, the
+Detection Engine will match against - if applicable:
+
+#. IPOnly rules (those where packet payload isn't inspected);
+#. Packet/payload related rules;
+#. Frame keywords;
+#. Application layer protocol transactions.
```

Comment:
This deserves a bit more explanation, as it has its own logic. Each parser has its own state machine, where it uses a per direction parsing state "progress". For each progress value keywords can be registered. E.g. http has a value "request line available" for which we registered keywords like http.uri, http.method, etc. If in the traffic we reach this state, the sigs with those keywords may be already evaluated, even if they have a lower prio than say a body inspecting sig.